### PR TITLE
REG: dt64 shift with integer fill_value

### LINKED ISSUE #31971: shift() broken for datetime64 when used with fill_value
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
from pandas.testing import assert_series_equal

dt_series = pd.Series((pd.to_datetime('2020-01-01'), pd.to_datetime('2020-01-02')))

actual_result = dt_series.shift(1, fill_value=0)
expected_result = pd.Series((pd.to_datetime('1970-01-01'), pd.to_datetime('2020-01-01')))

assert_series_equal(actual_result, expected_result)
```
#### Problem description

Above code snippet works on version 0.25.3 but breaks with 1.0.1.

```
print(actual_result)
0                      0
1    1577836800000000000
dtype: object

print(expected_result)
0   1970-01-01
1   2020-01-01
dtype: datetime64[ns]
```

```
AssertionError: Attributes of Series are different
Attribute "dtype" are different
[left]:  object
[right]: datetime64[ns]
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-72-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 1.0.1
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 45.2.0.post20200210
Cython           : None
pytest           : 5.3.2
hypothesis       : None
sphinx           : 2.3.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.1
IPython          : 7.11.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 0.15.1
pytables         : None
pytest           : 5.3.2
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.3.13
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #31971
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -745,6 +745,56 @@ def _from_factorized(cls, values, original):
...
+                FutureWarning,
+                stacklevel=7,
+            )
+            fill_value = new_fill
+
+        fill_value = self._unbox_scalar(fill_value)
+
+        new_values = self._data
+
+        # make sure array sent to np.roll is c_contiguous
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: This stuff is presumably covered by existing tests?

**Review by TomAugspurger — COMMENTED** [review]
Looks nice thanks.

Will need a release note in 1.0.2.

Do we have a policy on introducing warnings in bugfix releases? Technically this deprecation should have been done in 1.0, so maybe it's OK. Or the warning can wait till 1.1.

**Comment by TomAugspurger** [comment]
> Do we have a policy on introducing warnings in bugfix releases?

According to https://pandas.pydata.org/docs/development/policies.html 

> We will not introduce new deprecations in patch releases.

But perhaps it's OK here since this is fixing a bug (failure to deprecate something in 1.0). I don't have a strong preference.

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -745,6 +745,56 @@ def _from_factorized(cls, values, original):
     def _values_for_argsort(self):
         return self._data
 
+    def shift(self, periods=1, fill_value=None, axis=0):
+        if not self.size or periods == 0:
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: can you type things; I think we have an inherited doc-string?

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -745,6 +745,56 @@ def _from_factorized(cls, values, original):
...
+            return self.copy()
+
+        if is_valid_nat_for_dtype(fill_value, self.dtype):
+            fill_value = NaT
+        elif not isinstance(fill_value, self._recognized_scalars):
+            # only warn if we're not going to raise
+            if self._scalar_type is Period and lib.is_integer(fill_value):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: isinstance?

File: pandas/core/arrays/datetimelike.py [hunk 4]
```diff
@@ -745,6 +745,56 @@ def _from_factorized(cls, values, original):
...
+                # kludge for #31971
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: can you add a comment of what this is

File: pandas/core/arrays/datetimelike.py [hunk 5]
```diff
@@ -745,6 +745,56 @@ def _from_factorized(cls, values, original):
...
+        fill_value = self._unbox_scalar(fill_value)
+
+        new_values = self._data
+
+        # make sure array sent to np.roll is c_contiguous
+        f_ordered = new_values.flags.f_contiguous
+        if f_ordered:
+            new_values = new_values.T
+            axis = new_values.ndim - axis - 1
+
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: this is similar to the shift we have in block values yes? this should at some point be a routine in missing.py

Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: ideally doing it in this PR but if too hard to backport ok

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@jbrockmendel can you update doc-strings, otherwise lgtm.

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: yes.  this is copied near-verbatim from Block.shift (as jeff suggested, it should go somewhere else and be shared)

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: "is" is on purpose here

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: adding the docstring is easy, typing i tried and ended up punting on

File: pandas/core/arrays/datetimelike.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: Yah, i was thinking core.algos directory for algorithms-like functions that are specifically only ndarray/EA

**Comment by jbrockmendel** [comment]
updated+green; i think ive addressed the relevant comments

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks, let's see if this backports

**Comment by jreback** [comment]
@meeseeksdev backport 1.0.x

**Comment by jbrockmendel** [comment]
@TomAugspurger  did this get backported successfully or do i still need to do it manually?

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/pull/32647 backported it. I think the thing about the error is from the auto-backport and Jeff's manual request kicking in around the same time, so the second one errored since the backport was already created (maybe).

**Comment by jreback** [comment]
right i am sure not if things backport so i usuually kick the bot which i think is safe (it won’t do it twice)

but found that if it needs a manual backport it shows immediately which is good

**Pull request merged**